### DISCUSSION TITLE
4.1 Dev feedback

### DISCUSSION THREAD

**Discussion by grigi** [Blackymas/NSPanel_HA_Blueprint#1240] [General]
Just a place to do dev-for-4.1 feedback.

### 1. General feedback
* Since I last looked at it, you've made the build even smaller with esp-idf (1068105 bytes → 58%) which is awesome! :tada: 
* The panel _does_ feel faster with Nextion baud of 921600. Things are definitely snappier, and you get less of those subsystem blocked too long warning messages in the logs.

## Resolved issues
<details>
<summary>Recent regression</summary>

Somewhere between commit c838216c1ff65e4da17c8f7c2e7613eb1bc078b0 and dbc61f9ecb05a9571b99599862b5732241f8769c I've noticed the following regression:
* Using esp-idf, only including core, 921600 baud
* Panel initialisation now takes much much longer. It will now reset the screen twice before updating the display with the ESPHome version, but in the logs it repeats the version log entry about 12 times over the first minute:
```
[20:34:36][D][script.check_versions:2036]: ESPHome version: 4.1dev5
[20:34:36][D][script.check_versions:2037]: TFT version: 4.1dev5
[20:34:36][D][script.check_versions:2039]: Blueprint version: 4.1dev5
[20:34:36][D][script.check_versions:2048]: Framework: esp-idf
```
Will also see this during startup (not sure if relevant):
```
[20:35:25][W][nextion:331]: Nextion reported invalid instruction!
[20:35:25][E][nextion:274]: Nextion queue is empty!
```
After about 1 minute, right by the last (second) TFT flash I see this in the logs:
```
[21:04:44][W][nextion:719]: Received unknown event from nextion: 0x80
```
Then it will suddenly work, boot in and everything there works well (and fast).

Except that the logs is flooded with messages such as when interacting with anything:
```
[20:35:38][D][nextion:745]: Removing old queue type "NO_RESULT" name "send_command_printf"
[20:35:38][D][nextion:745]: Removing old queue type "NO_RESULT" name "backlight_brightness"
[20:35:38][D][nextion:745]: Removing old queue type "NO_RESULT" name "set_component_text"
[20:35:38][D][nextion:745]: Removing old queue type "NO_RESULT" name "set_component_text"
[20:35:38][D][nextion:745]: Removing old queue type "NO_RESULT" name "show_component"
[20:35:38][D][nextion:745]: Removing old queue type "NO_RESULT" name "send_command_printf"
[20:35:38][D][nextion:745]: Removing old queue type "NO_RESULT" name "set_component_text"
…
```

I wondered if this was to do with static IP, but disabling that made no change.
</details>

<details>
<summary>Logging via UART</summary>

I also noticed that ESPHome it is configured to send logs via UART0 at baud 115200. (Sending logs to the manually connected serial port via manual flashing wires). Disabling it with:
```yaml
logger:
  baud_rate: 0
```
would probably be a good idea to do by default as I found on my own ESPHome projects it helped marginally with CPU load.
You could have it enabled if someone includes "advanced".

</details>

**Comment by edwardtfn** [comment]
I'm investigating this "Removing old queue" thing... Nice that you could narrow the commits. I've done some change to the foreground/background script, it could be related...
Have you tried Arduino?

**Reply by grigi** [reply]
No, I haven't tried Arduino yet. I couldnt see anything between those commits that could generate those messages, Also the Nextion image didn't change in the mean time.

I'll try Arduino now.

**Reply by edwardtfn** [reply]
> I couldnt see anything between those commits that could generate those messages

Some of the changes happened on the Nextion component in ESPHome, as I'm moving there the changes made here at dev. It could be some change there... So many commits and discussions around this, I might have missed something. 😩

https://github.com/esphome/esphome/pulls?q=is%3Apr+nextion+

**Reply by edwardtfn** [reply]
Looks like the idf support will be included on ESPHome 2023.11, so I have to run to fix this before the release.

**Reply by grigi** [reply]
Wow, you've been doing LOTS of maintenance work on the Nextion component! :heart: 

Arduino build is 1215093 bytes 66.2% (still notably smaller than before :tada:)
Behaviour seems pretty much identical with Arduino. 

I'm NOT seeing these messages in the logs anymore:
```
[20:35:25][W][nextion:331]: Nextion reported invalid instruction!
[20:35:25][E][nextion:274]: Nextion queue is empty!
```
But since I disabled UART0 logging I also stopped seeing that in ESP-IDF
This might be coincidence though.

**Reply by edwardtfn** [reply]
> Wow, you've been doing LOTS of maintenance work on the Nextion component! ❤️

I've even made the changes to support idf v5, which will soon be supported by ESPHome. 😃 
But that fails with other components (API), so I'm not gonna suggest you guys to try that. 😉

**Reply by grigi** [reply]
Well, I'm available to do experimental build testing for the next few months before I'll have to put this panel into proper service (when moving).

**Reply by edwardtfn** [reply]
Let's take the item 1 from your feedback on #1241

**Comment by edwardtfn** [comment]
Logger via UART is now disable by default. 😉 
https://github.com/Blackymas/NSPanel_HA_Blueprint/wiki/(EN)-Customization#logger-via-uart

Thanks for the tip.

**Comment by edwardtfn** [comment]
Here you can see a preview of the release notes: https://github.com/Blackymas/NSPanel_HA_Blueprint/blob/dev/ReleaseNotes.md

**Reply by grigi** [reply]
That's an impressive collection so far. Completely missed the larger icons. I should try that.
I should also test out the media player stuff, but have no "smart" media player.
Or climate control thing.

:thinking: Hoes HA support fake sensors for integration testing?

**Reply by edwardtfn** [reply]
> Hoes HA support fake sensors for integration testing?

A lot of integrations supports "templates" entities. I use that a lot for my tests.
For Media Player however that might be a bit more complex, as you have to work quite a bit with templates to have something working... In other hands, a smart TV, even an old one, or an Alexa, etc, are pretty good as media players in Home Assistant.

**Comment by edwardtfn** [comment]
And here you can see what is in the backlog: https://github.com/Blackymas/NSPanel_HA_Blueprint/milestone/1

**Comment by edwardtfn** [comment]
Hi all,

Looks like we have a breaking change with ESPHome 2023.11 which will affect both the versions on `main` and `dev`, which will force me to stop any new feature to v4.1 now, prepare for that change and then release it.

I woukd love if you can run a last batch of tests and let me know about any issue so we can fix the critical ones before the release.

**Reply by edwardtfn** [reply]
More details here: https://github.com/Blackymas/NSPanel_HA_Blueprint/issues/1247

**Reply by grigi** [reply]
I build 4.1b1 and it works just fine on ESPHome 2023.10.6

I have to say the new TFT update mechanism is so much better. Less failures, much better performance, and now the partial uploads. Such an improvement!

**Reply by Doughboy68** [reply]
I updated my test panel to idf via serial and works fine.
I tried adding bluetooth proxy support  and reloaded via wireless, and my panel went unresponsive after the upload. Hard reboot just shows black screen.

I reflashed via serial without bluetooth proxy and is OK. Then I repeated the steps and same unresponsive panel.

I realize bluetooth proxy is not officially supported, but there is reference to release notes so I would look into this, and if there is an issue then advise to remove the reference in customizations.

Either I am doing something wrong in my config, or there is an issue with it now.


Other than than, I notice no issues so far. Initial panel boot and blueprint discovery is much faster now.

**Reply by edwardtfn** [reply]
I probably will go to a release tomorrow, as ESPHome 2023.11 will break the version on `main`, so I will remove the references to Bluetooth proxy for now and we come back to this later.

**Reply by edwardtfn** [reply]
Thanks again for running all those tests.

**Reply by edwardtfn** [reply]
And I believe the problem here is with ESP getting unstable when the memory is above certain levels. One thing you could try is to reference to `nspanel_esphome_core.yaml` only in your yaml. Remove the references to `nspanel_esphome.yaml` (which includes a reference to the TFT upload, which is very memory consumer) or `nspanel_esphome_advanced.yaml` (which includes `web_server`, also a memory demanding component).
We probably can shave the system a bit more (captive portal, etc), but I doubt that will make a significant difference.

**Reply by edwardtfn** [reply]
That combination will give the lower numbers I could get so far:
```log
RAM:   [=         ]   9.3% (used 30600 bytes from 327680 bytes)
Flash: [=====     ]  50.9% (used 934181 bytes from 1835008 bytes)
```

**Comment by Doughboy68** [comment]
> I build 4.1b1 and it works just fine on ESPHome 2023.10.6
> 
> I have to say the new TFT update mechanism is so much better. Less failures, much better performance, and now the partial uploads. Such an improvement!

Yes the new speed 921600 is very fast. But I noticed partial uploads for long time ever since I got a panel. Maybe one would not normally notice when upgrading major versions, but when I was testing lots on the last beta, I was constantly uploading new TFT and if there were minor changes it would start around 80% for example.

**Reply by edwardtfn** [reply]
That's right, the partial uploads are not new.

**Comment by Doughboy68** [comment]
I am not concerned about Bluetooth proxy really, but thought I’d give it a try therefore so just reporting my issue to avoid others having to take panels off wall and flash them via serial.

**Reply by Doughboy68** [reply]
I see there is a new blank_tft - just tested and works good.
I think procedure I will try to update my other four panels when 4.1 is released is:
copy new blank_tft to my files, rename it to current version in yaml so don't have to recompile.
Update TFT... since is blank won't take very long at 115200
Compile EspHome NSPanel with baud 921600 and install to NSPanel
Upload new tft version to my files.
Update TFT... should be fast at 921600

**Reply by edwardtfn** [reply]
About your plan:

> 1. Copy new blank_tft to my files, rename it to current version in yaml so don't have to recompile.
> 2. Update TFT... since is blank won't take very long at 115200
> 3. Compile EspHome NSPanel with baud 921600 and install to NSPanel
> 4. Upload new tft version to my files.
> 5. Update TFT... should be fast at 921600

After step 3, make sure to wait until ESPHome gets connected to Nextion before you go to step 4. Now you have an indication of ESPHome version on the `nspanel_blank.tft`.
It will fail uploading the new TFT until it gets connected to Nextion. This should take less than a minute.

**Comment by edwardtfn** [comment]
I'm wrapping up everything for the release. Please shout if you find anything critical.

**Reply by grigi** [reply]
I have to say the larger chips is a big deal for my mum-in-law. She asked me what those little pictures that look like doors mean.

She couldn't resolve the detail on the smaller ones.

**Reply by grigi** [reply]
Possibly the only constructive feedback I can give, is that the README on the home page is quite out of date.
In particular the pictures of how the panel screens look. The homepage and startup pages have changed a lot, There's the climate stuff and the media player, that are missing entirely.

**Comment by edwardtfn** [comment]
The new version is out: [v4.1 - Easier TFT transfe](https://github.com/Blackymas/NSPanel_HA_Blueprint/discussions/1255)

Thanks a lot for your support on this.
